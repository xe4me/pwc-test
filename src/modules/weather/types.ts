/*
 * @interface Forecast
 * This interface represents a day's weather attributes
 * More information about each field can be found here https://www.weatherbit.io/api/swaggerui/weather-api-v2#!/Daily32Historical32Weather32Data/get_history_daily_city_city_country_country
 * */
export interface Forecast {
  wind_cdir: string;
  pres: number;
  high_temp: number;
  wind_gust_spd: number;
  wind_spd: number;
  dewpt: number;
  weather: {
    icon: string;
    code: number;
    description: string;
  };
  wind_dir: number;
  precip: number;
  low_temp: number;
  max_temp: number;
  datetime: string;
  temp: number;
  min_temp: number;
}
/*
 * @interface CurrentWeatherApiResponse
 * This interfaces is a minimal version what the API actually returns, all the non-required fields are omitted for brevity
 * */
export interface CurrentWeatherApiResponse {
  city_name: string;
  data: Forecast[];
}

/*
 * @enum Unit
 * Two types of units, Metric and Scientific
 * */
export enum Unit {
  Metric = 'm',
  Scientific = 's',
}
