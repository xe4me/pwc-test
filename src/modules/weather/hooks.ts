import useSWR from 'swr';
import { createEndpoint } from 'utils/createEndpoint';
import { fetcher } from 'utils/fetcher';
import { CurrentWeatherApiResponse, Unit } from './types';

export const useCurrentWeatherData = ({
  unit,
  city,
}: {
  unit: Unit;
  city: string;
}) => {
  const { data, error } = useSWR<CurrentWeatherApiResponse | null>(
    createEndpoint(
      `/forecast/daily?city=${city}&country=australia&state=nsw&days=7&units=${unit}`
    ),
    fetcher,
    {}
  );
  return { data, error };
};
