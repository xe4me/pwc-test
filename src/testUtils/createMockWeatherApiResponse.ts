export const createMockWeatherApiResponse = (city_name: string) => ({
    city_name,
    data: [
      {
        wind_cdir: 'W',
        pres: 1014,
        high_temp: 20.2,
        wind_gust_spd: 5.9,
        wind_spd: 3.1,
        dewpt: 5.9,
        weather: {
          icon: 'c02d',
          code: 801,
          description: 'Few clouds',
        },
        wind_dir: 266,
        precip: 0,
        low_temp: 9.3,
        max_temp: 20.2,
        datetime: '2022-06-25',
        temp: 14.9,
        min_temp: 9.9,
      },
    ],
  });
