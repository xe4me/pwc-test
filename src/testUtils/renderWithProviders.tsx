import { render } from '@testing-library/react';
import { ReactElement } from 'react';
import { ThemeProvider } from '@mui/material';
import theme from '../theme';

export const renderWithProviders = (element: ReactElement) =>
  render(<ThemeProvider theme={theme}>{element}</ThemeProvider>);
