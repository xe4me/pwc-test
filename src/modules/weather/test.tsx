import React from 'react';
import { screen } from '@testing-library/react';
import { createMockWeatherApiResponse } from 'testUtils/createMockWeatherApiResponse';
import { renderWithProviders } from 'testUtils/renderWithProviders';
import { axe } from 'jest-axe';
import { Weather } from './index';
import * as useCurrentWeatherDataHook from './hooks';
import { CurrentWeatherApiResponse } from './types';

const mockWeatherHook = (
  data: CurrentWeatherApiResponse | null,
  error: any
) => {
  jest
    .spyOn(useCurrentWeatherDataHook, 'useCurrentWeatherData')
    .mockImplementation(() => ({
      data,
      error,
    }));
};

describe('<Weather>', () => {
  describe('Snapshot and accessibility', () => {
    it('should match with the snapshot', () => {
      const { baseElement } = renderWithProviders(<Weather />);
      expect(baseElement).toMatchSnapshot();
    });
    it('should pass accessibility tests', async () => {
      const html = renderWithProviders(<Weather />);
      expect(await axe(html.container)).toHaveNoViolations();
    });
  });
  describe('Data fetching behaviours', () => {
    describe('When api errors', () => {
      it('should render "failed to load"', () => {
        mockWeatherHook(null, 'Something went wrong');
        renderWithProviders(<Weather />);
        expect(screen.queryByText('failed to load')).toBeInTheDocument();
      });
    });
    describe('When api succeeds', () => {
      it('should NOT render "failed to load"', () => {
        mockWeatherHook(createMockWeatherApiResponse('Sydney'), null);
        renderWithProviders(<Weather />);
        expect(screen.queryByText('failed to load')).not.toBeInTheDocument();
      });
    });
  });
});
