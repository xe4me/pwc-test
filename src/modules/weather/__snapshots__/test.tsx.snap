// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Weather> Snapshot and accessibility should match with the snapshot 1`] = `
<body>
  <div>
    <div
      class="css-1reuysi"
    >
      <span
        aria-label="Weather data is loading ..."
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorSecondary css-1szo49z-MuiCircularProgress-root"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
  </div>
</body>
`;
